### PR TITLE
lemmy: fix ui commit hash

### PR DESCRIPTION
**Pull request by happysalada**
## Description of changes

@n3oney @Freakmiko when you have a moment, this is my understanding of the fix for the ui commit hash

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by happysalada** [comment]
@n3oney did you have time to test ?

**Comment by n3oney** [comment]
Sorry, I forgot. I also forgot how to do it :| let me quickly research

File: pkgs/servers/web-apps/lemmy/ui.nix [hunk 1]
```diff
@@ -62,6 +62,7 @@ mkYarnPackage {
   buildPhase = ''
     # Yarn writes cache directories etc to $HOME.
     export HOME=$PWD/yarn_home
+    export COMMIT_HASH=${src.rev}
```
Comment (n3oney) [inline pkgs/servers/web-apps/lemmy/ui.nix:65]: This doesn't work, as the COMMIT_HASH variable gets overridden in the package.json.

You'd have to either patch the package.json, or go the "fake git binary" route that I mentioned in the original PR.

**Review by n3oney — CHANGES_REQUESTED** [review]

File: pkgs/servers/web-apps/lemmy/ui.nix (hunk 1 above)
Comment (happysalada) [reply pkgs/servers/web-apps/lemmy/ui.nix]: @n3oney let me know if the new commit solves your issue then.

Comment (n3oney) [reply pkgs/servers/web-apps/lemmy/ui.nix]:
```
error: builder for '/nix/store/f6nf4ljp1il7ah4ilfwbzla00mi4hylx-lemmy-ui.drv' failed with exit code 1;
       last 8 log lines:
       > unpacking sources
       > unpacking source archive /nix/store/ixhd4ks8fw7jprbmr34g88r0phn79009-source
       > source root is source
       > patching sources
       > updateAutotoolsGnuConfigScriptsPhase
       > configuring
       > building
       > substitute(): ERROR: file 'package.json' does not exist
       For full logs, run 'nix log /nix/store/f6nf4ljp1il7ah4ilfwbzla00mi4hylx-lemmy-ui.drv'.
       ```

**Comment by n3oney** [comment]
I think you should be able to use something like this in place of the `./package.json` on line 52:

```nix
pkgs.stdenvNoCC.mkDerivation {
  src = ./package.json;
  name = "package.json";
  preferLocalBuild = true;
  allowSubstitutes = false;
  builder = pkgs.writeShellScript "substitute" ''
    source $stdenv/setup
    substitute "$src" "$out" --replace '$(git rev-parse --short HEAD)' "${src.rev}"

    eval "$postInstall"
  '';
}
```

**Review by n3oney — CHANGES_REQUESTED** [review]

**Comment by n3oney** [comment]
> I think you should be able to use something like this in place of the `./package.json` on line 52:

This doesn't seem to change anything and lemmy still builds without the commit hash... so I guess your `substituteInPlace` was more correct.

**Comment by n3oney** [comment]
Whoops @happysalada, turns out the solution is to simply move the substitution into the `patchPhase`.

This seems to build fine for me, although I can't test the functionality:

```nix
    patchPhase = ''
      substituteInPlace ./package.json \
        --replace '$(git rev-parse --short HEAD)' "${src.rev}" \
        --replace 'yarn clean' 'yarn --offline clean' \
        --replace 'yarn run rimraf dist' 'yarn --offline run rimraf dist'
    '';
```

The other 2 patches are simply to stop yarn from throwing these warnings:
`warning You don't appear to have an internet connection. Try the --offline flag to use the cache for registry queries.`

**Comment by n3oney** [comment]
Actually, nevermind, I did test it by simply running `nix run nodejs result/dist/js/server.js` and it works. 
image

**Comment by happysalada** [comment]
@n3oney thanks for the proposal.
Let me know if everything works as expected.

**Comment by n3oney** [comment]
`building lemmy-ui (buildPhase): $ webpack --env COMMIT_HASH=0.18.4 --mode=production` seems promising.....

aaand it works! https://lemmy.neoney.dev

**Pull request merged**